src/contexts/LoginContext.js

```diff
@@ -1,3 +1,61 @@
-import { createContext } from 'react';
+import React, { createContext, useEffect, useState, useContext } from 'react';
+import axios from 'axios';
 
-export const LoginContext = createContext(null);
+const { REACT_APP_BACKEND_URL: BACKEND_URL } = process.env
+const LOCAL_STORAGE_USER_ID_KEY = "userId"
+
+const LoginContext = createContext(null);
+
+const LoginConsumer = ({ children }) => {
+  const loginContextData = useContext(LoginContext)
+  if (!loginContextData) throw new Error('useLoginContext must be used within an LoginProvider')
+
+  return <LoginContext.Consumer>{children}</LoginContext.Consumer>
+}
+
+const LoginProvider = ({ children }) => {
+  const [userId, setUserId] = useState(localStorage.getItem(LOCAL_STORAGE_USER_ID_KEY))
+
+  const verifyLoginAndSetLocalStorage = async () => {
+    const resp = await axios.get(`${BACKEND_URL}/auth/whoami`)
+    const { userId } = resp.data
+    setUserId(userId)
+    localStorage.setItem(LOCAL_STORAGE_USER_ID_KEY, userId)
```

Comment:
The logic in lines 17, 23 for retrieving and storing `userId` into `localStorage` seems somewhat circular.

With this refactor, would it make sense to now, instead of storing `LOCAL_STORAGE_USER_ID_KEY` in `localStorage`, access the `userId` from the `LoginContext` in all our other components, e.g.

`const { userId } = useContext(LoginContext)`

Reply:
Re: the first point, if we don't store it in `localStorage`, we initialize `userId` in `LoginProvider` as `null` and that would cause the `ProtectedRoute` to send us to the login page if we close a tab and open a new one / on refresh (i.e. the login state does not get persisted). You can try this by setting the initialization value of the `useState` to be `null` or `''` instead of `localStorage.getItem(LOCAL_STORAGE_USER_ID_KEY)`.

Reply:
That's a valid point. I was thinking about how since`verifyLoginAndSetLocalStorage` will run anyway on refresh or on load, retrieving it from `localStorage` in line 17 is just a workaround from having the page flicker from `/` to `ProtectedRoute` on refresh. But I think the page flicker is not a good UI for users anyway, so happy to keep it in local storage.

Reply:
With the current PR implementation, if we didn't retrieve the `userId` from local storage, there wouldn't just be a page flicker - instead, the user will always be routed the following way (assuming that they land on a protected route, e.g. a workspace page):
- The user refreshes the browser window on a workspace page; `userId` would always be `null` since we don't store it in the browser local storage
- Since `verifyLoginAndSetLocalStorage` is an asynchronous network call, it takes a while to complete running. At this moment, the `<ProtectedRoute>` component is unable to find `userId` immediately and redirects the user to the `/` path
- `verifyLoginAndSetLocalStorage` completes successfully, and `userId` is stored in local storage.
- Since we are on `/`, `<RedirectIfLoggedInRoute>` kicks in and redirects the user to the `/sites` path